New article about changing password URL (#3799)
GoogleChrome/web.dev: src/site/content/en/blog/change-password-url/index.md

---

> title: Help users change passwords easily by adding a well-known URL for changing passwords
> subhead: Redirect a request to `/.well-known/change-password` to the change-passwords URL
> authors:
>   - agektmr
> date: 2020-08-31
> hero: hero.jpg
> alt: A drawing of a woman using OTP to log in to a web app.
> 
> description: |
>   By redirecting requests to /.well-known/change-password to the change password URL, you can let users update their passwords easier than before.
> 
> tags:
>   - blog # blog is a required tag for the article to show up in the blog.
>   - identity
>   - security

Reviewer: ```suggestion
title: Help users change passwords easily by adding a well-known URL for changing passwords
subhead: Redirect a request to `/.well-known/change-password` to the change-passwords URL
authors:
  - agektmr
date: 2020-08-31
hero: hero.jpg
alt: Neon lights shaped into various symbols that are commonly used in passwords.
description: |
  By redirecting requests to /.well-known/change-password to the change password URL, you can let users update their passwords easier than before.
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - identity
  - security
```

---

Set a redirect from `/.well-known/change-password` to the change password page
of your website. This will enable password managers to navigate your users
directly to that page.

## Introduction

As you may know, [passwords are not the best way to manage
accounts](https://security.googleblog.com/2017/11/new-research-understanding-root-cause.html).
Luckily, there are emerging technologies such as

> [WebAuthn](https://webauthn.io/) and techniques such as one time passwords that

Reviewer: ```suggestion
[WebAuthn](https://webauthn.io/) and techniques such as one-time passwords that
```

the correct input heuristically even if the website is not optimized for this
purpose. Web developers can help password managers by correctly [annotating HTML
input tags](https://web.dev/sign-in-form-best-practices/#new-password).

**Prevent phishing**: Because password managers remember where the password was
recorded, the password can be autofilled only at appropriate URLs, and not at
phishing websites.

**Generate strong and unique passwords**: Because strong and unique passwords
are directly generated and stored by the password manager, users don't have to
remember a single character of the password.

Generating and autofilling passwords using a password manager have already
served the web well, but considering their lifecycle, updating the passwords
whenever it's required is as important as generating and autofilling. To
properly leverage that, password managers are adding a new feature:

**Detect vulnerable passwords and suggest updating them**: Password managers can
detect passwords that are reused, analyze the entropy and weakness of them, and
even detect potentially leaked passwords or ones that are known to be unsafe

> from sources such as [https://haveibeenpwned.com/](https://haveibeenpwned.com/).

Reviewer: ```suggestion
from sources such as [Have I Been Pwned](https://haveibeenpwned.com/).
```

You can do two things to empower the user to update their password without
friction:

* If your change-password form needs the current password, add
  `autocomplete="current-password"` to the `<input>` tag to help the password
  manager autofill it.
* For the new password field (in many cases it's two fields to ensure that the
  user has entered the new password correctly), add
  `autocomplete="new-password"` to the `<input>` tag to help the password
  manager suggest a generated password.

Learn more at [Sign-in form best
practices](https://web.dev/sign-in-form-best-practices/#new-password).

## How it is used in real world

### Examples

Thanks to Apple Safari's
[implementation](https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes),
`/.well-known/change-password`, has already been available on some major
websites for a while:

* [Google](https://accounts.google.com/.well-known/change-password)
* [GitHub](https://github.com/.well-known/change-password)  
* [Facebook](https://www.facebook.com/.well-known/change-password)
* [Twitter](http://twitter.com/.well-known/change-password)

> * [Wordpress](https://wordpress.com/.well-known/change-password)

Reviewer: ```suggestion
* [WordPress](https://wordpress.com/.well-known/change-password)
```

Try them yourself and do the same for yours!

### Browser compatibility

A well-known URL for changing passwords has been [supported in Safari since
2019](https://webkit.org/blog/9170/safari-technology-preview-84-with-safari-13-features-is-now-available/).

> Chrome's password manager is starting to support it from version 86 onwards and
> other Chromium-based browsers may follow. [Firefox considers it worth

Reviewer: ```suggestion
Chrome's password manager is starting to support it from version 86 onwards
(which is scheduled for Stable release in late October 2020)
and other Chromium-based browsers may follow. [Firefox considers it worth
```

implementing](https://mozilla.github.io/standards-positions/#change-password-url),
but has not signalled that it plans to do so as of August 2020.

### Chrome's password manager behavior

Let's have a look at how Chrome's password manager treats vulnerable passwords.

Chrome's password manager is able to check for leaked passwords. By navigating

> to `chrome://settings/passwords` users can run Check passwords against stored

Reviewer: ```suggestion
to `chrome://settings/passwords` users can run **Check passwords** against stored
```

passwords, and see a list of passwords that are recommended for update.

<figure class="w-figure">

>   <img src="./check-passwords.png" width="1478" height="845" />

Reviewer: ```suggestion
  <img src="./check-passwords.png" width="1478" height="845" class="w-screenshot"/>
```

  <figcaption class="w-figcaption">

>     Chrome's "check passwords" functionality

Reviewer: ```suggestion
    Chrome's <b>Check passwords</b> functionality
```

  </figcaption>
</figure>

> By clicking the Change password button next to a password that is recommended to

Reviewer: ```suggestion
By clicking the <b>Change password</b> button next to a password that is recommended to
```

* A server-side-rendering website displays "Not found" when there is no content
  but with `200 OK`.
* A server-side-rendering website responds with `200 OK` when there is no
  content after redirecting to the "Not found" page.
* A single page app responds with the shell with `200 OK` and renders the "Not
  found" page on the client side when there is no content.

For these edge cases users will be forwarded to a "Not Found" page and that will
be a source of confusion.

That's why there's [a proposed standard
mechanism](https://wicg.github.io/change-password-url/response-code-reliability.html)
to determine whether the server is configured to respond with `404 Not Found`
when there is genuinely no content, by requesting a random page. Actually, the
URL is also reserved:
`/.well-known/resource-that-should-not-exist-whose-status-code-should-not-be-200`.
Chrome for example uses this URL path to determine whether it can expect a
proper change password URL from `/.well-known/change-password` in advance.

When you are deploying `/.well-known/change-password`, make sure that your
server returns `404 Not Found` for any non-existing contents.

{% endDetails %}

> ## Resources
> 
> If you have any feedback on the specification, please file an issue to [the spec
> repository](https://github.com/wicg/change-password-url/issues).
> 

Reviewer: ```suggestion
## Feedback

If you have any feedback on the specification, please file an issue to [the spec
repository](https://github.com/wicg/change-password-url/issues).

## Resources

```
